### PR TITLE
feat: convert project to typescript

### PR DESCRIPTION
**Pull request by juanpicado**
**Type:** refactor

**Description:**

Objective: **Convert the whole project to Typescript**

## Task

- [x] convert codebase 
- [x] No tsc errors
- [x] fix eslint errors
- [x] update docker image
- [x] update build stack to typescript
- [x] update @verdaccio/types 
- [x] Convert tests 
  - [x] Convert unit tests (WIP)
  - [x] Convert functional tests
  - [x] ~Convert e2e tests~ I skip this step, JS is fine for now.
- [x] add logging to local-storage (this help for convert test) https://github.com/verdaccio/local-storage/pull/172

## Context

I'm using the same setup we uses in all projects already converted:

* common babel configuration (https://github.com/verdaccio/babel-preset)  
* common eslint (https://github.com/verdaccio/eslint-config-verdaccio)
* common error handlers (https://github.com/verdaccio/commons-api)
* add types (https://github.com/verdaccio/flow-types)

I tried as much as possible to do not refactor anything not related with this migration.

Most likely unless is a critical or potential bug future tickets should be created if something does not look nice. Otherwise this will take for ever. I rely in all test now are passing, so I guess there are no regression issues. 

I plan to squash all the commits. This might affect current open PR, but I cannot wait any longer.

## What's next

- Fix all warnings
- Remove `@ts-ignore` as possible
- Remove `// @flow` or `@prettier` (those comments are harmless)
- Improve types (there are a lot of room for improvement here)
- Ideas 🤔 

Resolves https://github.com/verdaccio/verdaccio/issues/1166

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -29,15 +30,14 @@
...
     "handlebars": "4.1.2",
-    "http-errors": "1.7.3",
     "js-yaml": "3.13.1",
     "jsonwebtoken": "8.5.1",
     "kleur": "3.0.3",
-    "lodash": "4.17.11",
+    "lodash": "4.17.14",
     "lunr-mutable-indexes": "2.3.2",
-    "marked": "0.7.0",
+    "marked": "0.6.3",
```
Comment (sergiohgz) [inline package.json]: Is it any problem with this dep? A downgrade is done

File: src/api/endpoint/api/star.ts [hunk 2]
```diff
@@ -1,64 +1,64 @@
...
-            return users;
-        }, {});
-        storage.changePackage(name, { ...info, users}, req.body._rev, function(err) {
-          afterChangePackage(err);
-        });
-      },
-    });
-  };
-}
+// @flow
```
Comment (sergiohgz) [inline src/api/endpoint/api/star.ts]: Isn't this merge removing flow??

File: package.json (hunk 1 above)
Comment (juanpicado) [reply package.json]: Good catch

File: src/api/endpoint/api/stars.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
+/**
+ * @prettier
```
Comment (sergiohgz) [inline src/api/endpoint/api/stars.ts]: Should this pragma be removed?

File: src/lib/.eslintrc [hunk 4]
```diff
@@ -1,5 +1,6 @@
 {
```
Comment (sergiohgz) [inline src/lib/.eslintrc]: Is it necessary to have several eslintrc files across the project??

File: src/config/env.ts [hunk 5]
```diff
@@ -1,6 +1,5 @@
 /**
  * @prettier
```
Comment (sergiohgz) [inline src/config/env.ts]: Pragma idem

File: src/lib/bootstrap.ts [hunk 6]
```diff
@@ -1,21 +1,19 @@
 /**
  * @prettier
```
Comment (sergiohgz) [inline src/lib/bootstrap.ts]: Pragma idem

File: types/index.ts [hunk 7]
```diff
@@ -1,6 +1,7 @@
+/* eslint-disable @typescript-eslint/interface-name-prefix */
 // @flow
```
Comment (sergiohgz) [inline types/index.ts]: Flow pragma

File: src/lib/storage.ts [hunk 8]
```diff
@@ -139,37 +143,48 @@ class Storage implements IStorageHandler {
     // flow: should be IReadTarball
```
Comment (sergiohgz) [inline src/lib/storage.ts]: This comment don't apply now

File: src/lib/utils.ts [hunk 9]
```diff
@@ -1,6 +1,5 @@
 /**
  * @prettier
```
Comment (sergiohgz) [inline src/lib/utils.ts]: Pragma idem

File: src/api/endpoint/api/star.ts (hunk 2 above)
Comment (juanpicado) [reply src/api/endpoint/api/star.ts]: hehe ... there are some of them, the comment is harmless, I guess someone else can clean up that in a future PR :-)

File: src/api/endpoint/api/stars.ts (hunk 3 above)
Comment (juanpicado) [reply src/api/endpoint/api/stars.ts]: Actually it is one of my comments above in the What's next section.

File: src/lib/.eslintrc (hunk 4 above)
Comment (juanpicado) [reply src/lib/.eslintrc]: yes, actually some parts are less strict than others, for instance, `console.log` is allowed in some parts, or, actually I do not care about return type in unit test functions, for the other side, there was so many errors perhaps some `.eslintrc` file are just for display warnings and future clean up phase.

Comment (juanpicado) [reply src/lib/.eslintrc]: Those files basically override or extends the root configuration.

File: src/utils/user.ts [hunk 10]
```diff
@@ -33,7 +33,7 @@ export const GENERIC_AVATAR: string = `
 /**
  * Generate gravatar url from email address
  */
-export function generateGravatarUrl(email: string = '', online: boolean = true): string {
+export function generateGravatarUrl(email: string | void = '', online: boolean = true): string {
```
Comment (griffithtp) [inline src/utils/user.ts]: void or undefined?

File: src/lib/storage.ts [hunk 11]
```diff
@@ -16,26 +16,30 @@ import { checkPackageLocal, publishPackage, checkPackageRemote, cleanUpLinksRef,
...
-  localStorage: IStorage;
-  config: Config;
-  logger: Logger;
-  uplinks: ProxyList;
-  filters: Array<IPluginStorageFilter>;
+  public localStorage: IStorage;
+  public config: Config;
+  public logger: Logger;
+  public uplinks: ProxyList;
+  public filters: IPluginFilters;
```
Comment (griffithtp) [inline src/lib/storage.ts]: do these really need public scope?

File: src/lib/storage-utils.ts [hunk 12]
```diff
@@ -188,13 +220,14 @@ export function mergeUplinkTimeIntoLocal(localMetadata: Package, remoteMetadata:
...
+export function prepareSearchPackage(data: Package, time: unknown): any {
+  const listVersions: string[] = Object.keys(data.versions);
+  const versions: string[] = semverSort(listVersions);
+  const latest: string | undefined = data[DIST_TAGS] && data[DIST_TAGS].latest ? data[DIST_TAGS].latest : versions.pop();
 
-  if (data.versions[latest]) {
+  if (latest && data.versions[latest]) {
     const version: Version = data.versions[latest];
+    const versions: any = { [latest]: 'latest' };
     const pkg: any = {
```
Comment (griffithtp) [inline src/lib/storage-utils.ts]: how different is this from `Package` type?

File: src/api/endpoint/api/star.ts [hunk 13]
```diff
@@ -1,64 +1,64 @@
...
+
+import { USERS, HTTP_STATUS } from '../../../lib/constants';
+import {Response} from 'express';
+import {$RequestExtend, $NextFunctionVer, IStorageHandler} from '../../../../types';
+import _ from 'lodash';
```
[the remark this answers is not in the document]
Comment (juanpicado) [reply src/api/endpoint/api/star.ts]: Due the nature how lodash handle the imports (https://github.com/lodash/lodash/blob/master/package.json#L6) a bundle makes really hard to achieve this, you would need to go with `import xxxx from lodash/xxx' style, but that' is really verbose and the gain is minimal.

File: src/api/web/endpoint/user.ts [hunk 14]
```diff
@@ -0,0 +1,72 @@
+/**
+ * @prettier
+ * @flow
```
[the remark this answers is not in the document]
Comment (juanpicado) [reply src/api/web/endpoint/user.ts]: :-) read 'What's next' above please.

File: types/index.ts [hunk 15]
```diff
@@ -78,14 +90,13 @@ export type Profile = {
...
 
-export type $RequestExtend = $Request & {remote_user?: any}
-export type $ResponseExtend = $Response & {cookies?: any}
-export type $NextFunctionVer = NextFunction & mixed;
-export type $SidebarPackage = Package & {latest: mixed}
-
+export type $RequestExtend = Request & {remote_user?: any; log: Logger}
+export type $ResponseExtend = Response & {cookies?: any}
+export type $NextFunctionVer = NextFunction & any;
+export type $SidebarPackage = Package & {latest: any}
```
[the remark this answers is not in the document]
Comment (juanpicado) [reply types/index.ts]: I don't understand the part *reused in other libraries*. But I agree using Interfaces. I chose this style due was the way to bypass the issue, I did not try with Interfaces.

File: src/utils/user.ts (hunk 10 above)
Comment (juanpicado) [reply src/utils/user.ts]: 😊 it's not the same?

File: src/lib/storage.ts (hunk 11 above)
Comment (juanpicado) [reply src/lib/storage.ts]: Not in many, frankly at some point I did this mechanically, seems a reason for PR.

File: src/lib/storage-utils.ts (hunk 12 above)
Comment (juanpicado) [reply src/lib/storage-utils.ts]: Not 100% sure, this code is being used by `npm search` command. But that's something might change soon https://github.com/verdaccio/verdaccio/issues/310 so I would put it in the bottom of the priority queue.

**Comment by juanpicado** [comment]
Thanks @sergiohgz @griffithtp for the feedback. Just one more thing, many types were chosen based on research what previous contributors tried to achieve, and the abusing of coercion or mutability of objects were and are common rule. With this migration (and previously Flow) we have caught plenty of potential issues. However, there are a lot room for improvement, so future PR are very welcome 👍

**Pull request merged**